app/src/amazonWebview/java/org/mozilla/focus/webview/FirefoxAmazonWebView.kt

```diff
@@ -116,6 +116,21 @@ internal class FirefoxAmazonWebView(
         destroyDrawingCache()
         return outBitmap
     }
+
+    override fun scrollByClamped(vx: Int, vy: Int) {
+        // This is not a true clamp: it can only stop us from
+        // continuing to scroll if we've already overscrolled.
+        val scrollX = clampScroll(vx) { canScrollHorizontally(it) }
```

Comment:
I'm a little confused by this kotlinism - what do the curly braces mean? Does this has to do with the function being inlined (?) so the `canScrollHorizontally` isn't passed in as an argument within the parens (like `scroll`), but is rather passed in through curly braces?

Is there another way to pass this into the function? e.g., what if the ordering of `scroll` and `canScroll` were reversed, how would this be written?

Reply:
In Kotlin, if the last argument to a function is an anonymous function, you can move the anonymous function outside of the parenthesis. For example, these two invocations are equivalent:
```kotlin
val scrollX = clampScroll(vx) { canScrollHorizontally(it) }
val scrollX = clampScroll(vx, { canScrollHorizontally(it) })
```

---

> what if the ordering of scroll and canScroll were reversed, how would this be written?

If the anonymous function is not last, it must be passed inside the parenthesis:
```kotlin
val scrollX = clampScroll({canScrollHorizontally()it) }, vx)
```

> Does this has to do with the function being inlined (?)

No, inlining does not affect argument order. It primarily affects what your generated code looks like and [how `return` works](https://kotlinlang.org/docs/reference/inline-functions.html#non-local-returns).

---

The reason why Kotlin lets you move the anonymous function outside the parenthesis is to create Domain Specific Languages (or DSLs). For example, there could be an API to generate HTML with:
```kotlin
html {
    body {
        p { "Hey liuche!" }
    }
}
```

Instead of:
```kotlin
html.create(
    body.create(
        p.create("Hey liuche!")
    )
)
```

For sufficiently complex structures, it really helps to omit the redundancy.

In theory, we could also...
```kotlin
linearLayout {
    textView {
        withAttrs(
            layoutWidth = MATCH_PARENT,
            layoutHeight = MATCH_PARENT,
            text = "Hey liuche!"
        )
    }
}
```